### PR TITLE
Support partial guides (take 2)

### PR DESCRIPTION
**Pull request by burtbeckwith**

### CODE REVIEW

**Comment by burtbeckwith** [comment]
- Adds a new `publish` attribute to `GuideMetadata`: defaults to true for regular guides; set to false for partial/base guides
- Adds new `base` attribute to `GuideMetadata`: defaults to null; if set, indicates directory name of the base guide to copy before copying the current
- Adds new `external:` macro to include asciidoc from another guide or directory

With these changes, you can create a "base" guide, e.g. in `guides/some-guide-base`. The `metadata.json` would look like

```
{
  "publish": false,
  "slug": "some-guide-base",
  "authors": ["author1"],
  "tags": ["t1", "t2"],
  "category": "the category"
}
```

Here `publish` is false to indicate that this isn't a complete guide.

A "real" guide that's based on this one, e.g. `guides/some-guide-variant1` would have `metadata.json` like

```
{
  "base": "some-guide-base",
  "asciidoctor": "some-guide-variant1.adoc",
  "slug": "some-guide-variant1",
  "title": "some-guide-variant1",
  "intro": "Learn how to ...",
  "authors": ["author2"],
  "tags": ["t3", "t4"],
  "publicationDate": "2023-12-12",
  "apps": [
    {
      "name": "default",
      "features": ["graalvm","feature2"]
    }
  ]
}
```

Here `publish` is not set, so it defaults to true, and `base` is set to "some-guide-base" to indicate that this guide includes metadata and source files from `guides/some-guide-base`.

The `GuideMetadata` for `some-guide-variant1` will be merged with the `GuideMetadata` for `some-guide-base`, to allow setting defaults in the base guide.

`title` and `intro` default to the values in the main metadata but will use values from the base if not set.

`category`, `buildTools`, `languages`, `testFramework`, `skipGradleTests`, `skipMavenTests`, `minimumJavaVersion`, and `maximumJavaVersion` default to values from the base metadata but can be overridden in the main

The `authors` and `tags` lists will be combined from the base and main guide

`zipIncludes` and `apps` are not yet merged; they default to values from the base guide but can be overridden in the main

When generating the apps, existing files from the `guides/some-guide-base/groovy`, `guides/some-guide-base/java`, `guides/some-guide-base/kotlin`, and `guides/some-guide-base/src` directories will be copied first (after generating the initial Micronaut application), then existing files from the `guides/some-guide-variant1/groovy`, `guides/some-guide-variant1/java`, `guides/some-guide-variant1/kotlin`, and `guides/some-guide-variant1/src` directories.

The main guides' asciidoc file will use look like regular guide asciidoc, except that they'll include partial guide asciidoc files as needed with the `external:` macro, e.g.

```
<guide-specific asciidoc>

external:some-guide-base/some_include_from_base.adoc[]

<guide-specific asciidoc>

external:some-guide-base/some_other_include_from_base.adoc[]

<guide-specific asciidoc>
```

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy [hunk 1]
```diff
@@ -211,6 +220,26 @@ class GuideProjectGenerator implements AutoCloseable {
...
     }
 
+    private static void copyGuideSourceFiles(File inputDir, Path destinationPath,
+                                             String appName, String language,
+                                             boolean ignoreMissingDirectories) {
+
+        // look for a common 'src' directory shared by multiple languages and copy those files first
+        final String srcFolder = 'src'
+        Path srcPath = Paths.get(inputDir.absolutePath, appName, srcFolder)
+        if (srcPath.toFile().exists()) {
```
Comment (alvarosanchez) [inline buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]:
```suggestion
        if (Files.exists(srcPath)) {
```

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy [hunk 2]
```diff
@@ -211,6 +220,26 @@ class GuideProjectGenerator implements AutoCloseable {
...
+            Files.walkFileTree(srcPath, new CopyFileVisitor(Paths.get(destinationPath.toFile().path, srcFolder)))
```
Comment (alvarosanchez) [inline buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]:
```suggestion
            Files.walkFileTree(srcPath, new CopyFileVisitor(Paths.get(destinationPath, srcFolder)))
```

File: buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy [hunk 3]
```diff
@@ -211,6 +220,26 @@ class GuideProjectGenerator implements AutoCloseable {
...
+        }
+
+        Path sourcePath = Paths.get(inputDir.absolutePath, appName, language)
+        if (sourcePath.toFile().exists()) {
```
Comment (alvarosanchez) [inline buildSrc/src/main/groovy/io/micronaut/guides/GuideProjectGenerator.groovy]:
```suggestion
        if (Files.exists(sourcePath)) {
```

**Review by alvarosanchez — APPROVED** [review]

**Review by shingjo — APPROVED** [review]

**Comment by burtbeckwith** [comment]
@sdelamo OK to merge?

**Comment by burtbeckwith** [comment]
@sdelamo OK to merge?

**Comment by sdelamo** [comment]
I did not have the time to review. I will review from Master on Monday. I would have appreciated if you could have waited. 

Please, note there are other things waiting as well. e.g.  https://github.com/micronaut-projects/micronaut-guides/pull/674

**Pull request merged**